README.md
# Sensorium
Sensorium is a library for Makeblock's rj25 mainboards. It includes `mCore`, `Orion`, `Auriga` and  `Megapi`, `MegaPiPro`, and `Arduino` is also supported!
You can use it in browser and node enviroment.

# Install
```
npm install sensorium --save
```

# Usage
## Browser
Include the source `/browser/sensorium.js` in the root directory of project.
Actually you also need install [sensorium-server](https://www.npmjs.com/package/sensorium-server) as a proxy server.

```
npm install sensorium-server -g
```
Make sure your mainboards connecting to the your computer, and run `sensorium-server` with following command in your teminal:
```
ss
```
Then create a HTML page in the root directory, `index.html` for example, and type in the following script:
```
<script src="./browser/sensorium.js"></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.3/socket.io.js"></script>
<script>
  //firstly initialize
  var sensorium = new Sensorium();
  var mcore = sensorium.createMcore();

  //secondly set transport through `socket.io` like this:
  var socket = io.connect('http://localhost:8800');
  socket.on('sensor2web', function (data) {
      sensorium.doReceived(JSON.parse(data));
  });
  sensorium.setSender(function(buf) {
      socket.emit('web2sensor', { buf: JSON.stringify(buf), sensor: 'None' });
  });
</script>
<script>
  // set speed and run the motor.
  mcore.DcMotor(1).speed(200).run();

  // read ultrasonic sensor's value
  mcore.Ultrasonic(3).getData((val) => console.log(val));
</script>

```
## Nodejs
Firstly, install the dependencies:
```
npm install sensorium --save-dev
npm install serialport --save-dev
```

Touch a js file just like `test.js`, type in codes as follows:

```
var Sensorium = require("sensorium");
var SerialPort = require('serialport');
var sensorium = new Sensorium();

// Visit [serialport](https://www.npmjs.com/package/serialport) for detail usage
var serialPort = new SerialPort('/dev/ttyUSB0', { baudRate:115200 });

serialPort.on('open', function() {
  sensorium.setSender(function(buf) {
    serialPort.write(buf);
  });
  serialPort.on('data', function(buff) {
    sensorium.doReceived(buff);
  });
});

// optional mainboards are 'Mcore', 'Auriga', 'MegaPi', 'Orion', 'MegaPiPro', 'Arduino'
var auriga = sensorium.createAuriga();

// run DcMotor
auriga.DcMotor(1).speed(200).run();

// read ultrasonic sensor's value
auriga.Ultrasonic(3).getData().then(val => console.log(val));
```

Make sure your mainboards connecting to the your computer, and run srcipt with node:
```
node test.js
```

CHANGELOG.md
## Changelog
### [0.4.5] - 2018-08-20
#### Fix
-  fix bug

### [0.4.0] - 2018-05-23
#### Fix
-  fix bug of encoderMotor

### [0.3.9] - 2018-04-08
#### Add
-  Add method `setEmitMessage`,`getEmitMessage` for InfraredOnBoard module.

### [0.3.8] - 2018-04-04
#### Add
-  add two modules for mcore: `InfraredOnBoard`, `Infrared`

### [0.3.7] - 2018-04-03
#### Fixed
-  fix bug: missing some package when unpacking the received protocol.
### Changed
-  rename some modules mothod, such as `stop` to `setStop`, `reverse` to `setReverse`

### [0.3.6] - 2018-03-28
#### Fixed
-  fix bug: can't get right electronic module id when code was uglified.

### [0.3.5] - 2017-11-30
#### Fixed
-  fix the wrong spelling: `sensorium.doRecevied` => `sensorium.doReceived`.

### [0.3.3] - 2017-10-26
#### Fixed
-  fix bug of `sensorium.write` and `Control.pipe`

### [0.3.2] - 2017-09-29
#### Changed
-  unify the mainboard apis call as follows:
   - auriga.getVersion().getData().then((val) =>console.log(val))
   - auriga.setFirmwareMode().run()
   - auriga.readVoltage().getData().then((val) =>console.log(val))

### [0.3.1] - 2017-09-21
#### Fixed
-  fix bug: sensorium.SUPPORT instead of sensorium.getSupport()

### [0.3.0] - 2017-09-19
#### Changed
-  优化文件结构
-  Humiture 类将旧版本 API getHumidity, getTemperature 改为 readHumidity, readTemperature
-  EncoderMotorOnBoard 类将旧版本 API getSpeed, getAngle 改为 readSpeed, readAngle
-  FourLeds 类将旧版本类名 FourLed 改为 FourLeds
-  FourKeys 类将旧版本类名 FourKey 改为 FourKeys
-  所有的电子模块增加一个 protocol 方法，该方法为一个 getter 函数，以便提供获取协议的能力
-  规范所有模块提供 `run` 接口， 用于执行，并提供 `getData` 接口，用于读值。此外还提供了一些它们的封装接口

### [0.2.7] - 2017-09-12
#### Changed
-  优化文件结构
-  完善接口文档

## [0.2.6] - 2017-08-29
### Changed
-  sensorium 实例调用 setSender 设置传输方法
-  sensorium 实例调用 doReceived 将响应数据分发

### [0.2.5] - 2017-08-11
#### Fixed
-  Auriga Buzzer 模块无效的 bug
-  Auriga SmartServo 运动到相对位置、绝对位置等的 bug

### [0.2.4] - 2017-08-08
#### Added
- mcore 板载红外线传感器（读）接口 InfraredOnBoard
- 其他主控板外接红外线传感器（读）接口 Infrared
- LedMatrix 模块 4 种模式均增加 clear() 方法，以便能快捷的清除面板上显示的内容

#### Changed
- LedMatrix 模块 4 种模式显示内容的接口由原来的 ledMatrixData() 调整为 content()

#### Fixed
-  LedMatrix 表情面板显示字符不能工作的 bug

### [0.2.3] - 2017-08-08
#### Added
- Sensorium 实例新增 `send` 方法，该方法支持直接 write 协议
- Sensorium 库新增 API 文档，详情见 `./jsdoc` 目录

#### Fixed
-  Sorry for `async/await` throwing error. 已使用 babel-plugin-transform-runtime 转译修复此问题

### [0.2.2] - 2017-08-03
#### Added
- Sensorium 实例新增 `createMcore`,`createAuriga`,`createMegaPi`等方法创建对应的主控板实例

#### Changed
- 所有的读值接口，采用 Promise 回调的方式，先前版本通过传回调函数的方式被弃用
- getVerion 接口移至 Board 类实现，被各主控板继承
- LedMatrix 类重构，通过调用 `charMode`,`emotionMode`等方法进入不同的 ledMatrix 模式

### [0.2.1] - 2017-07-27
#### Added
- 提供了光线、温度、音量、led灯的板载 API

#### Changed
- 完善传各种参数给 API 时，参数少传、类型错传的容错机制，同时提供更加平滑的 warn 警告而程序不会被错误中断

### [0.2.1] - 2017-07-26
#### Added
- RgbLed 类中加了一个 rgb() 的接口，方便直接传入 hex color 值

#### Changed
- 优化了 Transport 类对通讯通道的初始化方式

#### Fixed
- 解决蓝牙返回值没有被监测到的 bug



src/settings.js
let
//超时时间
  OVERTIME = 1000,
  //自动超时
  AUTO_OVERTIME = 1200,
  FIRM_MODES = [0x00, 0x01, 0x02, 0x03, 0x04],
  SUPPORTLIST = ['Mcore', 'Auriga', 'MegaPi', 'Orion', 'Arduino', 'MegaPiPro'],
  FIRMWARE_ID = {
    0x06: 'Mcore',
    0x09: 'Auriga',
    0x0a: 'Orion',
    0x0e: 'MegaPi',
    0x0f: 'MegaPiPro',
    0x0d: "Arduino Mega 2560",
    
  },
  //音符到赫兹
  TONE_TO_HZ = {
    "B0": 31,
    "C1": 33,
    "D1": 37,
    "E1": 41,
    "F1": 44,
    "G1": 49,
    "A1": 55,
    "B1": 62,
    "C2": 65,
    "D2": 73,
    "E2": 82,
    "F2": 87,
    "G2": 98,
    "A2": 110,
    "B2": 123,
    "C3": 131,
    "D3": 147,
    "E3": 165,
    "F3": 175,
    "G3": 196,
    "A3": 220,
    "B3": 247,
    "C4": 262,
    "D4": 294,
    "E4": 330,
    "F4": 349,
    "G4": 392,
    "A4": 440,
    "B4": 494,
    "C5": 523,
    "D5": 587,
    "E5": 659,
    "F5": 698,
    "G5": 784,
    "A5": 880,
    "B5": 988,
    "C6": 1047,
    "D6": 1175,
    "E6": 1319,
    "F6": 1397,
    "G6": 1568,
    "A6": 1760,
    "B6": 1976,
    "C7": 2093,
    "D7": 2349,
    "E7": 2637,
    "F7": 2794,
    "G7": 3136,
    "A7": 3520,
    "B7": 3951,
    "C8": 4186,
    "D8": 4699
  },
  VERSION = Symbol(),
  MOVE_DIRECTION = ['FORWARD', 'BACKWARD', 'TURNLEFT', 'TURNRIGHT'],
  INFRARED_BUTTON = {
    'A':    69,
    'B':    70,
    'C':    71,
    'D':    68,
    'E':    67,
    'F':    13,
    '0':    22,
    '1':    12,
    '2':    24,
    '3':    94,
    '4':    8,
    '5':    28,
    '6':    90,
    '7':    66,
    '8':    82,
    '9':    74,
    'up':   64,
    'down': 25,
    'left': 7,
    'right':9,
    'set':  21
  };

export {
  OVERTIME,
  AUTO_OVERTIME,
  FIRM_MODES,
  SUPPORTLIST,
  FIRMWARE_ID,
  TONE_TO_HZ,
  MOVE_DIRECTION,
  VERSION,
  INFRARED_BUTTON
};